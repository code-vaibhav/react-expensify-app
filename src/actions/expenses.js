import database from "../firebase/firebase";

// ADD-EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      amount = 0,
      description = "",
      note = "",
      createdAt = 0
    } = expenseData

    const expense = {amount, description, note, createdAt};
    return database.ref("expenses").push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }));
      });
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({id} = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});