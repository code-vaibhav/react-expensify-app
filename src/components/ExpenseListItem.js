import React from "react";
import {Link} from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

export const ExpenseListItem = ({description, amount, createdAt, id}) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>Description: {description}</h3>
    </Link>
    <p>Amount: {numeral(amount/100).format("$0,0.00")}</p>
    <p>CreatedAt: {moment(createdAt).format("MMMM Do, YYYY")}</p>
  </div>
);

export default ExpenseListItem;