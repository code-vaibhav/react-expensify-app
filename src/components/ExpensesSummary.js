import React from "react";
import { connect } from "react-redux";
import getExpensesTotal from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";
import numeral from "numeral";

export const ExpensesSummary = ({expensesCount, expensesTotal}) => {
  const expenseWord = expensesCount === 1 ? "expense" : "expenses";
  const amount = numeral(expensesTotal/100).format("$0,0.00");
  return (
    <div>
      <h1>
        Viewing {expensesCount} {expenseWord} totalling {amount}
      </h1>
    </div> 
  );
}

const mapStateToProps = (state) => {
  const expenses = selectExpenses(state.expenses, {...state.filters});
  return {
    expensesCount: expenses.length,
    expensesTotal: getExpensesTotal(expenses)
  }
}

export default connect(mapStateToProps)(ExpensesSummary);