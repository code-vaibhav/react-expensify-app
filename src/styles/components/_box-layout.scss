.box-layout {
  background: url('https://www.pcb.its.dot.gov/StandardsTraining/modtransit12/ppt/mt12ppt67.jpg');
  background-repeat: no-repeat;
  background-size: cover;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;   
}

.box-layout__box {
  color: white;
  h1 {
    margin: 0 0 $m-size 0;
    line-height: 1;
  }
  background: fade-out($color: rgb(5, 5, 5), $amount: 0.4);
  padding: $l-size $m-size;
  border-radius: 3px;
  text-align: center;
  width: 25rem;
} 
