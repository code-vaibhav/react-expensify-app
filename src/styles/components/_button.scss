.button {
  color: white;
  background-color: $blue;
  display: inline-block;
  text-decoration: none;
  padding: $s-size;
  line-height: 1;
  font-size: $font-size-large;
  font-weight: 300;
  border: none;
}

.button--secondary {
  background-color: #888888;
}

.button--link {
  background: none;
}