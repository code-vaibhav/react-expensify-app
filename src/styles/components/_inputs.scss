.text-input {
  border: 1px solid #cacccd;
  font-size: $font-size-large;
  font-weight: 300;
  padding: $s-size;
  height: 46px;
}

.select {
  @extend .text-input;
}

.textarea {
  @extend .text-input;
  height: 10rem;
}