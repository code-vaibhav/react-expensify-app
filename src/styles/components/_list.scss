.list-header {
  display: flex;
  justify-content: space-between;
  background-color: $off-white;
  border: 1px solid darken($color: $off-white, $amount: 7%);
  color: $grey;
  padding: $s-size $m-size;
}

.list-body {
  margin-bottom: $xl-size;
}

.list-item {
  border: 1px solid darken($off-white, 7%);
  border-top: none;
  color: $dark-grey;
  text-decoration: none;
  display: flex;
  padding: $s-size;
  flex-direction: column;
  transition: background 0.3s ease;
  &:hover {
    background-color: $off-white;
  }

  @media (min-width: $desktop-breakpoint) {
    align-items: center;
    flex-direction: row;
    justify-content: space-between;
    padding: $m-size;
  }
}

.list-item--message {
  align-items: center;
  justify-content: center;
  padding: $m-size;
  color: $grey;
  &:hover {
    background: none;
  }
}

.list-item__title {
  margin: 0;
  word-break: break-all;
}

.list-item__sub-title {
  color: $grey;
  font-size: $font-size-small;

}

.list-item__data {
  margin: $s-size 0 0 0;
  @media (min-width: $desktop-breakpoint) {
    margin: 0;
    padding-left: $s-size;
  }
  ;
}