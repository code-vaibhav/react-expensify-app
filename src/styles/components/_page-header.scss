.page-header {
  background-color: $off-white;
  margin-bottom: $l-size;
  padding: $l-size 0;
}

.page-header__actions {
  margin-top: $m-size;
}

.page-header__inline {
  display: flex;
  flex-direction: column;

  
  @media (min-width: $desktop-breakpoint) {
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
  }
}

.page-header__title {
  font-weight: 300;
  margin: 0 0 $s-size 0;
  span {
    font-weight: bold;
  }
  @media (min-width: $desktop-breakpoint) {
    margin: 0;
  }
}