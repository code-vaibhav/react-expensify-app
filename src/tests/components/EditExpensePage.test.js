import React from "react";
import {shallow} from "enzyme";
import {EditExpensePage} from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let startEditExpense, startRemoveExpense, wrapper, history;

beforeEach(() => {
  startEditExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = {push: jest.fn()}
  wrapper = shallow(
    <EditExpensePage
      startEditExpense={startEditExpense} 
      startRemoveExpense={startRemoveExpense} 
      expense={expenses[0]}
      history={history}
    />);
});

test("Should render editExpensePage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("Should handle edit expense", () => {
  const expense = {
    amount: 185
  }
  wrapper.find("ExpenseForm").prop("onSubmit")(expense);
  expect(startEditExpense).toHaveBeenLastCalledWith(expenses[0].id, expense);
  expect(history.push).toHaveBeenLastCalledWith("/");
});

test("Should handle remove expense", () => {
  wrapper.find("button").simulate("click");
  expect(startRemoveExpense).toHaveBeenCalledWith(expenses[0].id);
  expect(history.push).toHaveBeenLastCalledWith("/");
});