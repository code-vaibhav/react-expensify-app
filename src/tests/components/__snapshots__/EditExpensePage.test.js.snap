// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render editExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container page-header__inline"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
      <button
        className="button button--secondary"
        onClick={[Function]}
      >
        Remove Expense
      </button>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "Gum",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
  </div>
</div>
`;
